server/queue/queue-activemq/src/main/java/org/apache/james/queue/activemq/metric/ActiveMQQueueStatistics.java

```diff
@@ -0,0 +1,155 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.activemq.metric;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.jms.JMSException;
+import javax.jms.MapMessage;
+
+import org.apache.james.metrics.api.GaugeRegistry;
+
+import com.google.common.util.concurrent.AtomicDouble;
+
+public class ActiveMQQueueStatistics implements ActiveMQStatistics {
+    protected static final String MEMORY_LIMIT = "memoryLimit";
+
+    protected static final String SIZE = "size";
+    protected static final String ENQUEUE_COUNT = "enqueueCount";
+    protected static final String DEQUEUE_COUNT = "dequeueCount";
+    protected static final String INFLIGHT_COUNT = "inflightCount";
+    protected static final String PRODUCER_COUNT = "producerCount";
+    protected static final String CONSUMER_COUNT = "consumerCount";
+    protected static final String EXPIRED_COUNT = "expiredCount";
+    protected static final String DISPATCH_COUNT = "dispatchCount";
+    protected static final String MESSAGES_CACHED = "messagesCached";
+
+    protected static final String MIN_ENQUEUE_TIME = "minEnqueueTime";
+    protected static final String MAX_ENQUEUE_TIME = "maxEnqueueTime";
+    protected static final String AVERAGE_ENQUEUE_TIME = "averageEnqueueTime";
+
+    protected static final String LAST_UPDATE = "lastUpdate";
+
+    private final AtomicLong memoryLimit = new AtomicLong();
+
+    private final AtomicLong size = new AtomicLong();
+    private final AtomicLong enqueueCount = new AtomicLong();
+    private final AtomicLong dequeueCount = new AtomicLong();
+    private final AtomicLong inflightCount = new AtomicLong();
+    private final AtomicLong producerCount = new AtomicLong();
+    private final AtomicLong consumerCount = new AtomicLong();
+    private final AtomicLong expiredCount = new AtomicLong();
+    private final AtomicLong dispatchCount = new AtomicLong();
+    private final AtomicLong messagesCached = new AtomicLong();
+
+    private final AtomicDouble minEnqueueTime = new AtomicDouble();
+    private final AtomicDouble maxEnqueueTime = new AtomicDouble();
+    private final AtomicDouble averageEnqueueTime = new AtomicDouble();
```

Comment:
If you want my opinion, maintaining ourselves a counter read by a gauge sounds overkill.

Dropwizard metrics comes with a `Histogram` that allows easily recording value.

This could be an idea of enhancement here.

Reply:
I see Dropwizard has such feature, but the metrics-api does not provide any interfaces yet. It only uses the histogram for timer metrics. Can you elaborate what you mean exactly?

Reply:
Yes sure. The overall feedback here is that it is (according to me) the metric stack job to store metrics state and that we do not need to store it ourselves.

I had a deeper look at dropwizard library. There is a handy object: `DefaultSettableGauge`...

```
        metricRegistry.register("", metric);
        metric.setValue(36L);
        metric.setValue(2L);
```

That would be very handy to me to design a little wrapper around this class in `GaugeRegistry`:

```
public interface GaugeRegistry {
    interface SettableGauge<T> {
        void setValue(T t);
    }

    <T> GaugeRegistry register(String name, Gauge<T> gauge);

    <T> SettableGauge<T> settableGauge(String name);
}
```

Then

```
public class DropWizardGaugeRegistry implements GaugeRegistry {
    private final MetricRegistry metricRegistry;

    @Inject
    public DropWizardGaugeRegistry(MetricRegistry metricRegistry) {
        this.metricRegistry = metricRegistry;
    }

    @Override
    public <T> GaugeRegistry register(String name, Gauge<T> gauge) {
        metricRegistry.gauge(name, () -> gauge::get);
        return this;
    }

    @PreDestroy
    public void shutDown() {
        metricRegistry.getGauges().keySet().forEach(metricRegistry::remove);
    }

    @Override
    public <T> SettableGauge<T> settableGauge(String name) {
        DefaultSettableGauge<T> gauge = new DefaultSettableGauge<>();
        metricRegistry.register(name, gauge);
        return gauge::setValue;
    }
}
```

Then in your AMQP metrics we could *directly* use these `SettableGauge` instead of the `Atomic*` fields...

What do you think about such a proposal?

Reply:
That's what I thought. The metrics-api needs some additional interfaces. I did not want to change the metrics-api in the first place, that's why I used the Atomic* fields.

Overall your proposal looks good to me and some boilerplate code will be removed. I have to check if `gauge::setValue` imposes any chance for blocking code. Since I use `Mono#timeout` to limit `fetchAndUpdate` of the statistics/metrics.

Should I apply your metrics-api change?

Reply:
>  I have to check if gauge::setValue imposes any chance for blocking code.

Blockhound will tell you but my bet is that no. Because dropwizrd relies on in-memory datastructures.

Reply:
> Should I apply your metrics-api change?

If you feel motivated to do this, sure, go for it!

Reply:
> If you feel motivated to do this, sure, go for it!

Ok, I will do so as part of this PR.

Reply:
done!